packages/backend/src/queryBuilder.test.ts

```diff
@@ -1038,9 +1051,126 @@ describe('applyLimitToSqlQuery', () => {
+        const sqlQuery = `
+            WITH subquery AS (
+                SELECT * FROM orders LIMIT 5
+            )
+            SELECT * FROM subquery LIMIT 10;
+        `;
+        const limit = 20;
+
+        // existing outer limit is 10 (less than 20), so return original query
+        const result = applyLimitToSqlQuery({ sqlQuery, limit });
+
+        expect(result).toBe(sqlQuery);
+    });
+
+    it('should apply new limit when existing limit is greater than the provided limit, even in subqueries', () => {
+        const sqlQuery = `
+            WITH subquery AS (
+                SELECT * FROM orders LIMIT 5
+            )
+            SELECT * FROM subquery LIMIT 25;
+        `;
+        const limit = 20;
+
+        const expectedQuery = `WITH user_sql AS (\nWITH subquery AS (\nSELECT * FROM orders\n)\nSELECT * FROM subquery\n) select * from user_sql limit ${limit}`;
+        const result = applyLimitToSqlQuery({ sqlQuery, limit });
+
+        expect(normalizeSQL(result)).toBe(normalizeSQL(expectedQuery));
+    });
+
+    it('should not remove LIMIT in table or field names', () => {
+        const sqlQuery = `
+            SELECT limit_column FROM limit_table WHERE limit_table.id = 10;
+        `;
+        const limit = 10;
+
+        // adjust the expected query to exclude the semicolon
+        const expectedQuery = `WITH user_sql AS (\n${sqlQuery
+            .trim()
+            .replace(/;$/, '')}\n) select * from user_sql limit ${limit}`;
+
+        const result = applyLimitToSqlQuery({ sqlQuery, limit });
+
+        expect(normalizeSQL(result)).toBe(normalizeSQL(expectedQuery));
+    });
+
+    it('should handle queries where LIMIT is inline and with a break line', () => {
+        const sqlQuery = `
+            SELECT * FROM users
+            LIMIT
+            10
+        `;
+        const limit = 5;
+
+        // existing limit is 10, which is greater than provided limit 5
+        const expectedQuery = `WITH user_sql AS (\nSELECT * FROM users\n) select * from user_sql limit ${limit}`;
+
+        const result = applyLimitToSqlQuery({ sqlQuery, limit });
+
+        expect(normalizeSQL(result)).toBe(normalizeSQL(expectedQuery));
+    });
+
+    it('should handle queries with LIMIT and OFFSET', () => {
+        const sqlQuery = `
+            SELECT * FROM users LIMIT 10 OFFSET 5;
+        `;
+        const limit = 20;
+
+        // existing limit is 10, which is less than provided limit 20
+        const result = applyLimitToSqlQuery({ sqlQuery, limit });
+
+        expect(result).toBe(sqlQuery);
+    });
+
+    it('should handle queries with LIMIT in strings', () => {
+        const sqlQuery = `
+            SELECT * FROM users WHERE name = 'LIMIT 10';
+        `;
+        const limit = 10;
+
+        // Adjust the expected query to exclude the semicolon
+        const expectedQuery = `WITH user_sql AS (\n${sqlQuery
+            .trim()
+            .replace(/;$/, '')}\n) select * from user_sql limit ${limit}`;
```

Comment:
If possible, hardcode the expected result. Otherwise, we might fall into a problem of coding the same logic that we are testing.